### PR TITLE
Fix bugs in the shared Clipboard code for WPF #11401 and #11402.

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Refactor ReadUtf8StringFromHGLOBAL for safer null-termination and error handling. Prevent double-release of IStream in TryGetIStreamData. Return null instead of corrupted data when GetData fails. Add comprehensive unit tests and new test mocks for edge cases. Update NativeMethods.txt with CLIPBRD_E_CANT_OPEN. Fixes issues [#11401](https://github.com/dotnet/winforms/issues/14308) and [#11402](https://github.com/dotnet/winforms/issues/14322) in the WPF repo (which also apply to WinForms as the code is shared now).
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/winforms/pull/14257)

### CODE REVIEW

File: src/System.Private.Windows.Core/tests/System.Private.Windows.Core.Tests/System/Private/Windows/Ole/NativeToManagedAdapterTests.cs [hunk 1]
```diff
@@ -181,4 +181,198 @@ public void ReadStringFromHGLOBAL_Terminator_ReturnsString(bool unicode)
...
+
+    [Fact]
+    public void ReadUtf8StringFromHGLOBAL_InvalidHGLOBAL_Throws()
+    {
+        Type type = typeof(Composition).GetFullNestedType("NativeToManagedAdapter");
+
+        Action action = () =>
+        {
+            string result = type.TestAccessor.Dynamic.ReadUtf8StringFromHGLOBAL(HGLOBAL.Null);
+        };
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Private.Windows.Core/tests/System.Private.Windows.Core.Tests/System/Private/Windows/Ole/NativeToManagedAdapterTests.cs:193]: For debugging purposes, this is useful

File: src/System.Private.Windows.Core/src/System/Private/Windows/Ole/Composition.NativeToManagedAdapter.cs [hunk 2]
```diff
@@ -226,7 +226,7 @@ private static unsafe string ReadStringFromHGLOBAL(HGLOBAL hglobal, bool unicode
             }
         }
 
-        private static unsafe string ReadUtf8StringFromHGLOBAL(HGLOBAL hglobal)
+        private static string ReadUtf8StringFromHGLOBAL(HGLOBAL hglobal)
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Private.Windows.Core/src/System/Private/Windows/Ole/Composition.NativeToManagedAdapter.cs:229]: It was duplicated from the class.

File: src/System.Private.Windows.Core/src/System/Private/Windows/Ole/Composition.NativeToManagedAdapter.cs [hunk 3]
```diff
@@ -237,17 +237,30 @@ private static unsafe string ReadUtf8StringFromHGLOBAL(HGLOBAL hglobal)
...
+                return Encoding.UTF8.GetString(bytes);
             }
             finally
             {
                 PInvokeCore.GlobalUnlock(hglobal);
             }
         }
 
-        private static unsafe string[]? ReadFileListFromHDROP(HDROP hdrop)
+        private static string[]? ReadFileListFromHDROP(HDROP hdrop)
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Private.Windows.Core/src/System/Private/Windows/Ole/Composition.NativeToManagedAdapter.cs:263]: Dupe from the class

File: src/System.Private.Windows.Core/src/System/Private/Windows/Ole/Composition.NativeToManagedAdapter.cs [hunk 4]
```diff
@@ -391,7 +410,7 @@ private static bool TryGetHGLOBALData<T>(
             return result;
         }
 
-        private static unsafe bool TryGetIStreamData<T>(
+        private static bool TryGetIStreamData<T>(
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Private.Windows.Core/src/System/Private/Windows/Ole/Composition.NativeToManagedAdapter.cs:413]: Dupe from the class.

**Comment by Shyam-Gupta** [comment]
Looks good.

**Comment by rampaa** [comment]
Does this PR need to be tagged with `servicing-consider` or something similar in order it to be serviced back to .NET 10?

**Comment by JeremyKuhne** [comment]
@rampaa: @Shyam-Gupta is going to determine how to manage this through the servicing process. It might be manual, separate PRs for each aspect of it.

**Comment by KlausLoeffelmann** [comment]
@JeremyKuhne, can you briefly explain what the main purpose of this PR was?
It is primarily a bug fix, is that correct?

If it _is_ a bug fix, is there any particular reason not to call it a bug fix?
You wrote

> This PR improves clipboard and OLE data handling with a focus on safety, correctness, and comprehensive testing. 

And you also wrote later

> Fixes issues https://github.com/dotnet/winforms/issues/14308 and https://github.com/dotnet/winforms/issues/14322 in the WPF repo (which also apply to WinForms as the code is shared now).

My question is: What does this PR do in addition to fixing the bugs, which would qualify for "improving clipboard and OLE data handling with a focus on safety, correctness, and comprehensive testing" which would go beyond correcting the respective mentioned bugs and adding unit test which would be added primarily in the context of the corrected bugs.

Further, for my own education:
When you write, it fixes the 2 referenced bugs in WPF, but the PR is changing files in the WinForms code base - how is the source code shared mechanically?

If we fix or change something in these respected code files here in the WinForms repo code base, is it then _automatically_ affecting the codebase in WPF (by some automatic replicating mechanism), or do the files need to be replicated (like, copied over) to WPF?

This is not urgent - I just need this for future reference.
I will also have Copilot assemble what kind of documentation status we have around this, and if it thinks we should add respective docs so folks would better know about the dependency between WPF and WinForms in this area - that is not a problem - because I will then trigger it to kick a PR for writing the docs with including a risk assessment off (meaning raising awareness that changing the one in WinForms means changing and potentially regressing the other).

@Shyam-Gupta, @merriemcgaw FYI.

Thanks!

**Comment by KlausLoeffelmann** [comment]
After consulting with copilot, I followed the advice and had it created 2 new issues and one PR with the docs.

@Shyam-Gupta and @JeremyKuhne:
Could you take a quick look as to the correctness of its approaches/conclusions, and provide feedback about kicking of the 2nd pr, which would add info about the shared nature of the clipboard code to the source files?

@JeremyKuhne, should there are already respective docs or comments in the source, and both Copilot and I should have missed them, could you please point me to them? In that case, I would quickly have copilot amend that info, and have it taken those for all the created items into account.

Thanks!

**Comment by rampaa** [comment]
@Shyam-Gupta Will there be no backport for the other regression we are experiencing https://github.com/dotnet/winforms/issues/14322?

**Comment by Shyam-Gupta** [comment]
> @Shyam-Gupta Will there be no backport for the other regression we are experiencing #14322?

I am still investigating that issue on priority, and the plan is to backport it once we have the fix. In our internal testing, we found that the changes in this PR do not address that issue.
